pkg/plugins/runtime/k8s/controllers/gatewayapi/http_route_controller.go

```diff
@@ -0,0 +1,184 @@
+)
+
+// HTTPRouteReconciler reconciles a GatewayAPI object into Kuma-native objects
+type HTTPRouteReconciler struct {
+	kube_client.Client
+	Reader kube_client.Reader
+	Log    logr.Logger
+
+	Scheme          *kube_runtime.Scheme
+	Converter       k8s_common.Converter
+	SystemNamespace string
+	ResourceManager manager.ResourceManager
+}
+
+// Reconcile handles transforming a gateway-api HTTPRoute into a Kuma
+// GatewayRoute and managing the status of the gateway-api objects.
+func (r *HTTPRouteReconciler) Reconcile(ctx context.Context, req kube_ctrl.Request) (kube_ctrl.Result, error) {
+	httpRoute := &gatewayapi.HTTPRoute{}
+	if err := r.Get(ctx, req.NamespacedName, httpRoute); err != nil {
+		if kube_apierrs.IsNotFound(err) {
+			err = r.ResourceManager.Delete(ctx, core_mesh.NewGatewayRouteResource(), store.DeleteByKey(req.Name, req.Name))
+			if store.IsResourceNotFound(err) {
+				err = nil
+			}
+		}
+
+		return kube_ctrl.Result{}, err
+	}
+
+	// TODO check that gateways exist etc
+	// TODO set status on gapi resources
+
+	resource := core_mesh.NewGatewayRouteResource()
+
+	coreName := util_k8s.K8sNamespacedNameToCoreName(httpRoute.Name, httpRoute.Namespace)
+	mesh := k8s_util.MeshFor(httpRoute)
+
+	if err := r.ResourceManager.Get(ctx, resource, store.GetByKey(coreName, mesh)); err != nil {
+		if !store.IsResourceNotFound(err) {
+			return kube_ctrl.Result{}, err
+		}
+
+		resource.Spec, err = r.gapiToKumaRoute(httpRoute.Namespace, httpRoute)
+		if err != nil {
+			return kube_ctrl.Result{}, err
+		}
+
+		resource.SetMeta(&k8s.KubernetesMetaAdapter{ObjectMeta: metav1.ObjectMeta{
+			Name:      httpRoute.Name,
+			Namespace: httpRoute.Namespace,
+		}, Mesh: mesh})
+
+		err := r.ResourceManager.Create(ctx, resource, store.CreateByKey(coreName, mesh))
+
+		return kube_ctrl.Result{}, errors.Wrap(err, "could not create GatewayRoute resource")
+	}
+
+	var err error
+
+	resource.Spec, err = r.gapiToKumaRoute(httpRoute.Namespace, httpRoute)
+	if err != nil {
+		return kube_ctrl.Result{}, err
+	}
+
+	err = r.ResourceManager.Update(ctx, resource, store.ModifiedAt(core.Now()))
+
+	return kube_ctrl.Result{}, errors.Wrap(err, "could not update GatewayRoute resource")
+}
+
+func (r *HTTPRouteReconciler) gapiToKumaRoute(namespace string, route *gatewayapi.HTTPRoute) (*mesh_proto.GatewayRoute, error) {
+	var selectors []*mesh_proto.Selector
+
+	// Convert GAPI parent refs into Kuma tag matchers
+	for _, ref := range route.Spec.ParentRefs {
+		namespace := route.Namespace
+		if ns := ref.Namespace; ns != nil {
+			namespace = string(*ns)
+		}
+
+		match := map[string]string{
+			mesh_proto.ServiceTag: serviceTagForGateway(string(ref.Name), namespace),
+		}
+		if sn := ref.SectionName; sn != nil {
+			match[mesh_proto.ListenerTag] = string(*sn)
+		}
+		// TODO if sn == nil, what does it mean to reference the entire resource, all
+		// listeners? check this works
+
+		selectors = append(selectors, &mesh_proto.Selector{
+			Match: match,
+		})
```

Comment:
I wonder if we should always generate selectors with hostname matches? Every route would have a selector to match hostname "*", but then we also append selectors for each hostname in the route. I think this is redundant since the gateway generator has to deal with it anyway for unmanaged tag selectors, but it might make it more obvious in this case.

:shrug:

Reply:
Hmm then we should only add a `"*"` hostname if we _don't_ have any hostnames in the `HTTPRoute`, right?

Reply:
Yeh let's pass on this, at least for now.